spyder/api/widgets.py

```diff
@@ -0,0 +1,1176 @@
+    @Slot()
+    def create_window(self):
+        """
+        Create a QMainWindow instance containing this SpyderWidget.
+        """
+        # Wigdets
+        self.windowwidget = window = SpyderWindowWidget(self)
+
+        # Wigdet setup
+        window.setWindowIcon(self.get_icon())
+        window.setCentralWidget(self)
+        window.setWindowTitle(self.get_title())
+        window.resize(self.size())
+
+        if self.dockwidget:
+            self.dockwidget.setFloating(False)
+            self.dockwidget.setVisible(False)
+
+        self.set_ancestor(window)
+        self._update_actions()
+        window.show()
+
+        # Signals
+        # TODO:
+        # self.sig_refresh_requested.emit()
+
+    @Slot()
+    def close_window(self):
+        """
+        Close QMainWindow instance that contains this SpyderWidget.
+        """
+        if self.windowwidget is not None:
+            # Fixes spyder-ide/spyder#10704
+            self.__unsafe_window = self.windowwidget
+            self.__unsafe_window.deleteLater()
+            self.windowwidget.close()
+            self.windowwidget = None
+
+            # These actions can appear disabled when 'Dock' action is pressed
+            self.undock_action.setDisabled(False)
+            self.close_action.setDisabled(False)
+
+        if self.dockwidget is not None:
+            # TODO: need a signal!
+            # self.set_ancestor(self.main)
+            self.dockwidget.setWidget(self)
+            self.dockwidget.setVisible(True)
+            self.dockwidget.raise_()
+            self._update_actions()
+            # self.sig_switch_to_plugin_requested.emit()
+
+    def change_visibility(self, enable, force_focus=None):
+        """
+        Dock widget visibility has changed.
+        """
+        if self.dockwidget is None:
+            return
+
+        is_visible = not self.is_visible
+
+        # TODO: handle maximized!
+
+        if enable:
+            # Avoid double trigger of visibility change
+            self.dockwidget.blockSignals(True)
+            self.dockwidget.raise_()
+            self.dockwidget.blockSignals(False)
+
+        raise_and_focus = getattr(self, 'RAISE_AND_FOCUS', None)
+
+        if force_focus is None:
+            if raise_and_focus or not is_visible:
+                focus_widget = self.get_focus_widget()
+                if focus_widget:
+                    focus_widget.setFocus()
+        elif force_focus is True:
+            focus_widget = self.get_focus_widget()
+            if focus_widget:
+                focus_widget.setFocus()
+        elif force_focus is False:
+            pass
+
+        self.is_visible = enable
+
+        # TODO: Handle this option
+        # if getattr(self, 'DISABLE_ACTIONS_WHEN_HIDDEN', None):
+        #     toggle_actions(self.get_actions(), visible)
+        #     toggle_actions(self._plugin_actions, visible)
+
+        # self.is_visible = enable and visible
```

Comment:
>Please fix this. As you can see in the last line here, enable and visible were two different things in the old API. I don't understand why you're changing that. Also, that could have subtle but serious implications for several panes.

The logic in the old API was not great, otherwise this major refactoring would not have been needed. I have not need to use this code in the testing I made.

So please test any of the over 10 PRs available with different plugins to ensure things work,.